homeassistant/components/intent.py

```diff
@@ -0,0 +1,178 @@
+"""Module to coordinate user intentions."""
+import asyncio
+import logging
+
+import voluptuous as vol
+
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+
+
+DOMAIN = 'intent'
+_LOGGER = logging.getLogger(__name__)
+
+SLOT_SCHEMA = vol.Schema({
+    vol.Optional('type'): str,
+})
+
+SPEECH_TYPE_PLAIN = 'plain'
+SPEECH_TYPE_SSML = 'ssml'
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup intent."""
+    hass.intent = IntentRegistry(hass)
```

Comment:
You're so right @andrey-git.

This is what I'm going to rewrite it to:

```
from homeassistant.components import intent

intent.register(hass, 'AddItemShoppingList', handler)
```

Then inside `register` the logic will be like this:

- If intent component loaded, register with intent registry via hass.data (it's in a method, so internal).
- If intent component is not loaded, store handler in `hass.data`. If it is setup, read all already stored handlers and register them. Because we're async, we know there is only 1 thread running and so we don't have to be afraid of race conditions 💃